### PR TITLE
config/opal_check_os_flavors.m4: fix -Wimplicit-int in OPAL_CHECK_OS_...

### PR DESCRIPTION
**Pull request by thesamesam**
…FLAVOR_SPECIFIC

Clang 16 makes -Wimplicit-intan error by default.

This one is harmless but it's very noisy, especially because it's used
a bunch in subconfigures. These warnings becoming errors can, especially in configure,
lead to miscompiled or misconfigured software, so distributions are starting
to look for instances of them in config.log, etc.

In this case, an error is supposed to fire (because it's checking platform
name) but it's easier to use the preprocessor directive '#error' so that
we don't trigger a -Wimplicit-int warning/error instead.

Fixes errors like:
```
error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Werror,-Wimplicit-int]
```

For more information, see LWN.net [0] or LLVM's Discourse [1], Gentoo's wiki [2]
or the (new) c-std-porting mailing list [3].

[0] https://lwn.net/Articles/913505/
[1] https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213
[2] https://wiki.gentoo.org/wiki/Modern_C_porting
[3] hosted at lists.linux.dev.

Signed-off-by: Sam James <sam@gentoo.org>

### CODE REVIEW

File: ompi/mca/op/avx/configure.m4 [hunk 1]
```diff
@@ -315,15 +315,15 @@ AC_DEFUN([MCA_ompi_op_avx_CONFIG],[
                        [$op_sse3_support],
                        [SSE3 supported in the current build])
     AM_CONDITIONAL([MCA_BUILD_ompi_op_has_avx512_support],
-                   [test "$op_avx512_support" == "1"])
+                   [test "$op_avx512_support" = "1"])
```
Comment (bwbarrett) [inline ompi/mca/op/avx/configure.m4:318]: This patch doesn't seem to have anything to do with other commits in this PR.

File: config/opal_check_os_flavors.m4 [hunk 2]
```diff
@@ -22,7 +22,7 @@ AC_DEFUN([OPAL_CHECK_OS_FLAVOR_SPECIFIC],
     AC_MSG_CHECKING([$1])
     AC_COMPILE_IFELSE([AC_LANG_PROGRAM(
      [[#ifndef $1
-      error: this is not $1
+      #error: this is not $1
```
Comment (bwbarrett) [inline config/opal_check_os_flavors.m4:25]: I don't understand why this change is the right change.  For reasons documented at https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/C-Compiler.html, using #error in an autoconf test is a no-no.  THe autoconf tests themselves use "choke me" instead of our nice pretty print, but I have difficulty seeing how ours can cause more problems than what AC does in general.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: ompi/mca/op/avx/configure.m4 (hunk 1 above)
Comment (thesamesam) [reply ompi/mca/op/avx/configure.m4:318]: Sure, I'll spit it out. I didn't do it prematurely given it's trivial, but it's no problem.

File: config/opal_setup_cc.m4 [hunk 3]
```diff
@@ -365,7 +365,7 @@ AC_DEFUN([OPAL_SETUP_CC],[
         [opal_cv_cc_supports___builtin_prefetch],
         [AC_LINK_IFELSE([AC_LANG_PROGRAM([],
           [int ptr;
-           __builtin_prefetch(&ptr,0,0);])],
+           int __builtin_prefetch(&ptr,0,0);])],
```
Comment (jsquyres) [inline config/opal_setup_cc.m4:368]: Can you clarify what this change is for?  From my reading, it changes the body of `main()` in the test from:
```c
int ptr;
__builtin_prefetch(&ptr,0,0);
```
to
```c
int ptr;
int __builtin_prefetch(&ptr,0,0);
```
which doesn't appear to be valid C.

File: config/opal_check_os_flavors.m4 (hunk 2 above)
Comment (thesamesam) [reply config/opal_check_os_flavors.m4:25]: Thanks, I didn't realise that was the autoconf guidance.

Let me explain then you can tell me what you think the best approach is:
- In general, we're looking for new Clang 16 (and GCC 14) breakage which is painful if it's in config.log
- Easiest way is grepping for -Wimplicit-int, -Wimplicit-function-declaration, etc
- The current state of the macro leads to a bunch of false positives:
```
conftest.c:84:7: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
      error: this is not __OpenBSD__
      ^
      int
conftest.c:84:12: error: expected ';' after top level declarator
      error: this is not __OpenBSD__
           ^
           ;
conftest.c:84:14: error: unknown type name 'this'
      error: this is not __OpenBSD__
             ^
conftest.c:84:21: error: expected ';' after top level declarator
      error: this is not __OpenBSD__
                    ^
                    ;
4 errors generated.
configure:12189: $? = 1
```

This isn't a common occurrence and it's the first time so far in the general porting effort that I've come across this construct.

If you insist on keeping it as it is, that's fine, it just makes life a lot easier when going through literally 1000s of these packages and looking for problems. It'll save someone else looking at it and realising it's actually fine in this case.

What do you think?

Comment (bwbarrett) [reply config/opal_check_os_flavors.m4:25]: Why are you not tripping over the same problem for the "Choke me" that is in every configure script out there that checks for GNU compiler and therefore runs this code:

```
int
main ()
{
#ifndef __GNUC__
       choke me
#endif

  ;
  return 0
```

Comment (bwbarrett) [reply config/opal_check_os_flavors.m4:25]: I think trying to grep that out of config.log is a path to madness.  But I have no objection to a patch that fits with Autoconf's guidance.  Not following Autoconf's guidance there seems like you're going to be generally chasing your tail with Autoconf macros themselves.

File: config/opal_setup_cc.m4 (hunk 3 above)
Comment (thesamesam) [reply config/opal_setup_cc.m4:368]: Sorry, that's my fault for hurrying. That's definitely dodgy. I got distracted doing something else (I'll file a separate PR for that) and was a bit autopilot by then.

But apparently I can't reproduce this bit anyway, sorry for the noise.

File: config/opal_check_os_flavors.m4 (hunk 2 above)
Comment (thesamesam) [reply config/opal_check_os_flavors.m4]: There's not really an alternative for what we're doing (see https://lwn.net/Articles/913505/). Surprisingly the autoconf macros themselves have been mostly fine, although they're going to make a new release for a few nits shortly. But I won't bore you with the details.

How's the latest thing I pushed?

File: ompi/mca/op/avx/configure.m4 (hunk 1 above)
Comment (thesamesam) [reply ompi/mca/op/avx/configure.m4]: Split into https://github.com/open-mpi/ompi/pull/11083.

**Review by bwbarrett — APPROVED** [review]

**Comment by jsquyres** [comment]
@thesamesam Can you cherry pick to the relevant branches?  I'm assuming this needs to go to v5.0.x, and probably v4.1.x.  I wouldn't worry about v4.0.x, though -- that branch is getting quite stale and is unlikely to see another release.

**Comment by rhc54** [comment]
I went ahead and ported this change to PMIx and PRRTE, but I'm wondering if we shouldn't restart work on OAC, perhaps making this the next one to bring across? I'm willing to help, but pretty much would just copy/rename what we have - so if you want a major revamp of an m4 (like what happened with the last one I tried to just copy/rename across), I'm not going to be able to help. Don't want to create a bunch of work for @bwbarrett
